app/models/company.rb

```diff
@@ -0,0 +1,38 @@
+class Company < ApplicationRecord
+  belongs_to :owner, class_name: 'User'
+  has_many :parent_shares, -> {
+                             where(active: true)
+                           }, class_name: 'Share', foreign_key: 'child_id', dependent: :destroy, inverse_of: :child
+  has_many :child_shares, -> {
+                            where(active: true)
+                          }, class_name: 'Share', foreign_key: 'parent_id', dependent: :destroy, inverse_of: :parent
+
+  def owning_users
+    return User.where(id: owner_id) if parent_shares.empty?
+
+    User.where(id: top_owning_companies.pluck(:owner_id))
+  end
+
+  def top_owning_companies
```

Comment:
@ulferts there must be a more elegant (and likely more efficient) way to do this. 

I thought about using `closure_tree`, `awesome_nested_set`, or one of the other tree gems that you have in the project, but I couldn't easily wrap my head around how to make them work with interleaving models.

In order to make this work, I'd think of making some sort of abstraction (e.g. `Owneships`), with a polymorphic association for Shares/Companies... However, this is the kind of thing that I'd first like to discuss with a sparring partner such as yourself, before changing the relationships implied in the supplied spec.